bitcoin/contrib/test.sh

```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/usr/bin/env bash
```

Comment:
If you don't want to use `env` then we have to use `sh` because that's the only shell that Nix users have in `/bin` or `/usr/bin`.

Reply:
That's weird, what's the reason? Why can't they install `/bin/bash` like every other distribution?

Reply:
Because Nix doesn't "install" anything in the normal FHS paths because those are unversioned. Instead everything is stored in the Nix store with paths like `/nix/store/f2ibj8np6dyddf3kq68dkcvwfr6bmi9n-bash-interactive-5.2-p15/bin/bash` and "installed" by being symlinked into the correct location.

For /etc and some other directories it will symlink stuff into the FHS locations so that things appear normal (other than all your config files being read-only even to root). But for /bin it doesn't do this, presumably so that different users can have different sets/versions of binaries "installed".

Reply:
Yes, they have FS virtualization for steam :). But it's only used for a couple random things.

For the specific case of scripts bundled with normal software, Nix has a rule that'll rewrite the shebang lines. I could use this in my local CI. But people just running NixOS trying to run the shellscripts will get burned by it.

Reply:
Sounds like they'll stay "geeky OS" with this kind of approach. :(

Anyway, OK, I'm willing to accept env, but I would really like to have some version detection. Like if it finds that bash is older than let's say 4.4.12 (super conservative choice) it'll exit with an error message. Also maybe we should check things like `cargo` and `jq` being available upfront but that's fine to do separately.

Reply:
I'm happy with all those suggestions.

Note that even without `env` we have had problems with this in the past, where CI boxes were running bash 4 but everyone else in the world had been using bash 5 for years.

And I'm not sure "ergonomics of running shell scripts" is going to be the difference between a mainstream and a geeky OS ;).

Reply:
I've added a command to check for used commands. Also I've printed the version of `bash` being run. I do not have the motivation to work out how to grep the version output and assert its greater than 4 right now. I am well and truly tired of this PR.

Reply:
Lol, fine not to grep and assert on it. But the next time we have a "CI sort order is mysteriously different" bug we'll appreciate having it in the logs.

If you were less sick of this PR I'd suggest you print out the current locale as well :P.

Reply:
OK, hopefully version being printed will help.

> we have had problems with this in the past, where CI boxes were running bash 4 but everyone else in the world had been using bash 5 for years.

Wow! Good to know.

>  "ergonomics of running shell scripts"

Doesn't it apply to all programs? If not why exclude shell scripts?

> If you were less sick of this PR I'd suggest you print out the current locale as well

Or rather whole env.